<title>JSON - JavaScript Object Notation</title>

<script id="JSON format - serialization">

    // const student = {
    //     firstName: "Adam",
    //     lastName: "Foobar",
    //     isGraduate: true,
    //     friends: ["Beth", "Cecil", "Dorthy"],
    //     grades: {
    //         math: 5,
    //         physics: 4,
    //         chemistry: 4.5,
    //     },
    //     detentions: null,
    //     sayHi() {
    //         console.log("Hello there!");
    //     }
    // };

    // console.log(typeof student, student);
    // console.log(student.toString());


    // const serializedJson = JSON.stringify(student);
    // console.log(typeof serializedJson, serializedJson);

    // const serializableStudent = {
    //     ...student,
    //     toString() {
    //         return JSON.stringify(this);
    //     }
    // };

    // const studentJson = serializableStudent.toString();
    // console.log(typeof studentJson, studentJson);

</script>

<script id="JSON parsing - deserialization">

    // const parseAndLog = (serialized) => {
    //     const deserialized = JSON.parse(serialized);
    //     console.log(typeof deserialized, deserialized);
    //     return deserialized;
    // };

    // // parseAndLog("[3,2,1"); /* Oops */
    // const arr = parseAndLog("[3,2,1]");

    // const obj = parseAndLog(`{ "username": "Kuba" }`);

    // arr.push(obj);
    // console.log(arr);

</script>

<script id="Handling errors - try/catch">

    // const parseAndLogSafely = (serialized) => {
    //     try {
    //         const deserialized = JSON.parse(serialized);
    //         console.log(typeof deserialized, deserialized);
    //         return deserialized;
    //     } catch (error) {
    //         console.log("Failed to parse data.");
    //         console.log({ error });
    //         console.error(error);

    //         // console.log(error.message);
    //         // console.log(error.stack);
    //     }
    // };

    // parseAndLogSafely("[3,2,1"); /* Oops */
    // parseAndLogSafely("[3,2,1]"); // git gud



    // throw new Error("Some Error"); /* uncaught */
    // console.log("I will not run.");

    // try {
    //     throw new Error("Some Error");
    // } catch (error) {
    //     // console.log(error);
    //     // throw new Error("Some custom Error");

    //     console.log("Ooopsie.");
    //     throw error;
    // }

    // console.log("This won't run after the unhandled error!");


    

    // const add = (x, y) => {
    //     if (typeof x !== "number" || typeof y !== "number") {
    //         throw new Error(`Add arguments must be a number. x=${typeof x}, y=${typeof y}`);
    //     }
    //     return x + y;
    // };

    // const requireNumber = (value) => {
    //     const valueType = typeof value;
    //     if (valueType !== "number") {
    //         throw new Error(`Value must be a number. Got "${valueType}" of "${value}" instead.`);
    //     }
    //     return value;
    // };

    // const sum = (a, b) => requireNumber(a) + requireNumber(b);

    // console.log(add(40, "2")); /* faulty code */
    // console.log(add(40, 2));

    // console.log(sum(40, "2")); /* faulty code */
    // console.log(sum(40, 2));

</script>

<script id="Runtime errors vs Syntax/Compile* time errors">

    /* Syntax error */
    // led smthing = 42; /* typo */
    // let smthing = 42;
    // console.log(smthing);

    /* Runtime error */
    // console.log(JSON.parse("[42"));

</script>
